packages/arb-bridge-eth/contracts/arch/Marshaling.sol

```diff
@@ -0,0 +1,255 @@
+// SPDX-License-Identifier: Apache-2.0
+
+/*
+ * Copyright 2019-2020, Offchain Labs, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+pragma solidity ^0.5.11;
+
+import "./Value.sol";
+import "./Hashing.sol";
+
+import "../libraries/BytesLib.sol";
+
+library Marshaling {
+    using BytesLib for bytes;
+    using Value for Value.Data;
+
+    function deserializeHashPreImage(bytes memory data, uint256 startOffset)
+        internal
+        pure
+        returns (
+            uint256, // offset
+            Value.Data memory
+        )
+    {
+        uint256 size;
+        require(
+            data.length >= startOffset && data.length - startOffset >= 64,
+            "to short"
+        );
+        bytes32 hashData = data.toBytes32(startOffset);
+        startOffset += 32;
+        (startOffset, size) = deserializeInt(data, startOffset);
+        return (startOffset, Value.newTuplePreImage(hashData, size));
+    }
+
+    function deserializeInt(bytes memory data, uint256 startOffset)
+        internal
+        pure
+        returns (
+            uint256, // offset
+            uint256 // val
+        )
+    {
+        require(
+            data.length >= startOffset && data.length - startOffset >= 32,
+            "too short"
+        );
+        return (startOffset + 32, data.toUint(startOffset));
+    }
+
+    function deserializeCheckedInt(bytes memory data, uint256 startOffset)
+        internal
+        pure
+        returns (
+            bool, // valid
+            uint256, // offset
+            uint256 // val
+        )
+    {
+        uint256 totalLength = data.length;
+        if (
+            totalLength < startOffset ||
```

Comment:
what's the reasoning for at times failing a transaction and returning false at other times?

Reply:
By default I switched as much as I could failing instead of returning a bool since it's easier to deal with. However we still have a couple places where failure of parsing has to not revert the transaction. Both `unmarshalOutgoingMessage` and `bytestackToBytes` have to be able to fail without reverting.